client/lib/catch-js-errors/index.js

```diff
@@ -1,81 +1,121 @@
-			} else if ( assignment <= 0.02 ) {
-				localStorage.setItem( 'log-errors', 'analytics' );
-				errorLogger = 'analytics';
-			} else if ( assignment <= 0.03 ) {
-				localStorage.setItem( 'log-errors', 'external' );
-				//Prep the stage up for Google Stackdriver or other service experiment
-			} else {
-				localStorage.setItem( 'log-errors', 'false' );
 			}
-		}
 
-		if ( errorLogger === 'true' ) {
-			//Fix up any outstanding old settings
-			localStorage.setItem( 'log-errors', 'rest-api' );
-			errorLogger = 'rest-api';
+			if ( errorLogger === 'rest-api' ) {
+				// set up handler to POST errors
+				TraceKit.report.subscribe( errorReport => {
+					const error = {
+						message: errorReport.message,
+						url: document.location.href,
+						trace: errorReport.stack,
+					};
+
+					this.diagnose();
+					this.sendToApi( Object.assign( error, this.diagnosticData ) );
+				} );
+			} else if ( errorLogger === 'analytics' ) {
+				TraceKit.report.subscribe( function gaApiLogger( errorReport ) {
+					if ( typeof window.ga === 'function' ) {
+						window.ga(
+							'send',
+							'event',
+							'JS Error',
+							'URL: ' + document.location.href,
+							errorReport.message + ' ' + JSON.stringify( errorReport.stack ),
+							0,
+							{ nonInteraction: true }
+						);
+					}
+				} );
+			}
 		}
+	}
 
-		if ( errorLogger === 'rest-api' ) {
-			// set up handler to POST errors
-			TraceKit.report.subscribe( function restApiLogger( errorReport ) {
-				const error = {
-					message: errorReport.message,
-					url: document.location.href,
-					trace: errorReport.stack,
-				};
+	saveNewPath( newPath ) {
+		this.diagnosticData.extra.previousPaths.unshift( this.diagnosticData.path );
+		this.diagnosticData.extra.previousPaths = this.diagnosticData.extra.previousPaths.slice( 0, 4 );
+		this.diagnosticData.path = newPath;
+	}
 
-				if ( window.navigator && window.navigator.userAgent ) {
-					error.browser = window.navigator.userAgent;
-				}
+	saveDiagnosticReducer( data ) {
+		this.diagnosticReducers.push( data );
+	}
 
-				sendToApi( error );
-			} );
-		} else if ( errorLogger === 'analytics' ) {
-			TraceKit.report.subscribe( function gaApiLogger( errorReport ) {
-				if ( typeof window.ga === 'function' ) {
-					window.ga(
-						'send',
-						'event',
-						'JS Error',
-						'URL: ' + document.location.href,
-						errorReport.message + ' ' + JSON.stringify( errorReport.stack ),
-						0,
-						{ nonInteraction: true }
-					);
-				}
-			} );
+	saveDiagnosticData( data ) {
+		if ( typeof data.extra === 'object' ) {
+			this.saveExtraData( data );
+		} else {
+			Object.assign( this.diagnosticData, data );
 		}
 	}
-}() );
+
+	saveExtraData( data ) {
+		Object.assign( this.diagnosticData.extra, data );
+	}
+
+	diagnose() {
+		this.diagnosticReducers.forEach( diagnosticReducer => {
+			try {
+				this.saveDiagnosticData( diagnosticReducer() );
+			} catch ( e ) {
+				console.warn( 'diagnostic', this.diagnosticData );
```

Comment:
Maybe also add a flag to diagnostic data, so we know that something bad happened when we tried to get extra info.